Upgrade Preact Version + Test Refactor

swarajd opened the pull request:
Here's what's going on in this PR:

- The project has been started from a scratch preact boilerplate. The non-preact project dependencies (i.e. cytoscape, popper, tippy) remain mostly intact in terms of versions. 
- The tests have been modified to not run through node.js. Previously, they relied on the `fs` node module to load test files, but now I have slightly altered the tests to use `jest-raw-loader` to load the test files for the relevant tests and now the test suite runs using pure js. 
- `diagram.js` has been slightly refactored to get rid of the warning messages in the console that yell about certain elements not having the `fname` attribute to pull data from in the css. 

This is quite the large PR, so feel free to put it through the gauntlet in terms of testing and scrutiny to make sure it doesn't cause any unwanted changes. 

I tested this locally and it seemed to work based on my interaction with the example chart, but some other, lesser known/tested aspects could be broken due to package versioning changes and whatnot.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/capitalone/architecture-viewer?pullRequest=11) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/capitalone/architecture-viewer?pullRequest=11) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/capitalone/architecture-viewer?pullRequest=11) it.</sub>

daj commented on package.json:
```diff
@@ -4,64 +4,50 @@
   "version": "1.0.0",
   "license": "Apache-2.0",
   "scripts": {
-    "start": "if-env NODE_ENV=production && npm run -s serve || npm run -s dev",
```
The [Preact 10 docs](https://preactjs.com/guide/v10/getting-started) suggest the convention is to use `npm run dev`, so I can update the README to replace `npm start` with `npm run dev` (which worked fine when I tested it).

swarajd replied:
Good catch, I went ahead and replaced it in the README myself.

daj commented:
The `npm run build` step isn't working for me, any ideas?
```
$ npm run build

> architecture-viewer@1.0.0 build /Users/dj/Code/architecture-viewer-swaraj
> preact build

✖ ERROR Error: TypeError: Cannot destructure property 'plugin' of 'helpers.getPluginsByName(...)[0]' as it is undefined.
    at default (/Users/dj/Code/architecture-viewer-swaraj/preact.config.js:24:15)
    at module.exports (/Users/dj/Code/architecture-viewer-swaraj/node_modules/preact-cli/lib/lib/webpack/transform-config.js:117:10)
    at async prodBuild (/Users/dj/Code/architecture-viewer-swaraj/node_modules/preact-cli/lib/lib/webpack/run-webpack.js:87:2)
    at async module.exports (/Users/dj/Code/architecture-viewer-swaraj/node_modules/preact-cli/lib/commands/build.js:30:14)
    at module.exports (/Users/dj/Code/architecture-viewer-swaraj/node_modules/preact-cli/lib/lib/webpack/transform-config.js:128:10)
    at async prodBuild (/Users/dj/Code/architecture-viewer-swaraj/node_modules/preact-cli/lib/lib/webpack/run-webpack.js:87:2)
    at async module.exports (/Users/dj/Code/architecture-viewer-swaraj/node_modules/preact-cli/lib/commands/build.js:30:14)
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! architecture-viewer@1.0.0 build: `preact build`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the architecture-viewer@1.0.0 build script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/dj/.npm/_logs/2020-08-05T02_25_49_598Z-debug.log
```
My config:
```
node v12.18.3 (stable)
npm  v6.14.6
```
I also tried with `node v14.7.0 (latest)`, but the error seemed the same.

The other steps all seem to work fine in my testing so far! 👏

daj commented on package.json:
```diff
@@ -4,64 +4,50 @@
+    "enzyme-adapter-preact-pure": "^2.0.0",
+    "eslint": "^6.0.1",
+    "eslint-config-preact": "^1.1.0",
     "identity-obj-proxy": "^3.0.0",
```
I think we probably still want the license-checker.
```suggestion
    "identity-obj-proxy": "^3.0.0",
    "license-checker": "^20.2.0",
```

daj commented on package.json:
```diff
@@ -4,64 +4,50 @@
+    "build": "preact build",
+    "serve": "sirv build --port 8080 --cors --single",
     "dev": "preact watch",
     "lint": "eslint src",
```
I think we probably still want the license-checker. 
```suggestion
    "lint": "eslint src",
    "license": "license-checker --csv --production > licenses.csv"
```
I tested that this still works after introducing this line and the dependency further below.
```
$ npm run license

> architecture-viewer@1.0.0 license /Users/dj/Code/architecture-viewer-swaraj
> license-checker --csv --production > licenses.csv

$ cat licenses.csv 
"module name","license","repository"
"ajv@6.12.3","MIT","https://github.com/ajv-validator/ajv"
"architecture-viewer@1.0.0","UNLICENSED",""
"cytoscape-dagre@2.2.2","MIT","https://github.com/cytoscape/cytoscape.js-dagre"
"cytoscape-popper@1.0.7","MIT","https://github.com/cytoscape/cytoscape.js-popper"
"cytoscape@3.15.2","MIT","https://github.com/cytoscape/cytoscape.js"
"dagre@0.8.5","MIT","https://github.com/dagrejs/dagre"
"fast-deep-equal@3.1.3","MIT","https://github.com/epoberezkin/fast-deep-equal"
"fast-json-stable-stringify@2.1.0","MIT","https://github.com/epoberezkin/fast-json-stable-stringify"
"graphlib@2.1.8","MIT","https://github.com/dagrejs/graphlib"
"heap@0.2.6","PSF","https://github.com/qiao/heap.js"
"json-schema-traverse@0.4.1","MIT","https://github.com/epoberezkin/json-schema-traverse"
"lodash.debounce@4.0.8","MIT","https://github.com/lodash/lodash"
"lodash@4.17.19","MIT","https://github.com/lodash/lodash"
"popper.js@1.16.1","MIT","https://github.com/FezVrasta/popper.js"
"preact-render-to-string@5.1.10","MIT","https://github.com/developit/preact-render-to-string"
"preact-router@3.2.1","MIT","https://github.com/developit/preact-router"
"preact@10.4.6","MIT","https://github.com/preactjs/preact"
"pretty-format@3.8.0","MIT","https://github.com/thejameskyle/pretty-format"
"punycode@2.1.1","MIT","https://github.com/bestiejs/punycode.js"
"tippy.js@2.5.4","MIT","https://github.com/atomiks/tippyjs"
"uri-js@4.2.2","BSD-2-Clause","https://github.com/garycourt/uri-js"
```

daj commented on package.json:
```diff
@@ -4,64 +4,50 @@
-  "eslintIgnore": [
-    "build/*"
-  ],
   "devDependencies": {
```
I think we still want the gulp dependencies for the `deploy` task (I had to bump to the latest gulp version to get it to run - the older version gave me `ReferenceError: primordials is not defined`).
```suggestion
  "devDependencies": {
    "gulp": "^4.0.2",
    "gulp-gh-pages": "^0.5.4",
```
After adding these dependencies back in, the `deploy` task seemed to run (though this is blocked by `npm run build` not working for me yet):
```
$ npm run deploy

> architecture-viewer@1.0.0 deploy /Users/dj/Code/architecture-viewer-swaraj
> gulp deploy

[22:52:45] Using gulpfile ~/Code/architecture-viewer-swaraj/gulpfile.js
[22:52:45] Starting 'deploy'...
[22:52:45] [gh-pages] No files in the stream.
[22:52:45] Finished 'deploy' after 17 ms
```

daj commented on package.json:
```diff
@@ -4,64 +4,50 @@
+    "enzyme": "^3.10.0",
+    "enzyme-adapter-preact-pure": "^2.0.0",
+    "eslint": "^6.0.1",
+    "eslint-config-preact": "^1.1.0",
```
I think we still want the gulp dependencies for the `deploy` task (I had to bump to the latest gulp version to get it to run - the older version gave me `ReferenceError: primordials is not defined`).
```suggestion
    "eslint-config-preact": "^1.1.0",
    "gulp": "^4.0.2",
    "gulp-gh-pages": "^0.5.4",
```
After adding these dependencies back in, the `deploy` task seemed to run (though this is blocked by `npm run build` not working for me yet):
```
$ npm run deploy

> architecture-viewer@1.0.0 deploy /Users/dj/Code/architecture-viewer-swaraj
> gulp deploy

[22:52:45] Using gulpfile ~/Code/architecture-viewer-swaraj/gulpfile.js
[22:52:45] Starting 'deploy'...
[22:52:45] [gh-pages] No files in the stream.
[22:52:45] Finished 'deploy' after 17 ms
```

daj commented on package.json:
```diff
@@ -4,64 +4,50 @@
-    "test": "jest ./tests",
-    "deploy": "gulp deploy",
-    "license": "license-checker --csv --production > licenses.csv"
+    "test": "jest"
```
I think we still want the gulp dependencies for the deploy task.
```suggestion
    "deploy": "gulp deploy",
    "test": "jest"
```

swarajd commented:
Regarding `preact build` you'll probably have to do `npm run build -- --no-prerender` for it to work as the app uses DOM apis which prevent it from being able to be pre-rendered. 2b08d4bc7920e6ed6aff274b1c1b890f7809fab9 should fix the initial error but later on you'll have to build using `--no-prerender` anyways.

swarajd commented:
I also committed most of the changes you suggested, lmk if there's anything else that could use some tweaking.

daj commented on package.json:
```diff
@@ -4,64 +4,55 @@
-    "start": "if-env NODE_ENV=production && npm run -s serve || npm run -s dev",
-    "build": "preact build --no-prerender",
-    "serve": "preact build --no-prerender && preact serve",
+    "build": "preact build",
```
As you mentioned, this works for me:
```suggestion
    "build": "preact build --no-prerender",
```
Without it, I see this error when running `npm run build`:
```
ReferenceError: window is not defined
method: render
at: /Users/dj/Code/architecture-viewer-swaraj/src/components/app.js:197:16

Source code:

	}
	render() {
		let repoUrl = window.location.href.replace("pages/", "");
		if (this.state.graphData && this.state.stepData) {
			return (
				<div id="container">

This is most likely caused by using DOM or Web APIs.
Pre-render runs in node and has no access to globals available in browsers.

Consider wrapping code producing error in: 'if (typeof window !== "undefined") { ... }'

Alternatively use 'preact build --no-prerender' to disable prerendering.

See https://github.com/developit/preact-cli#pre-rendering for further information.npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! architecture-viewer@1.0.0 build: `preact build`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the architecture-viewer@1.0.0 build script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/dj/.npm/_logs/2020-08-06T02_05_14_313Z-debug.log
```

daj commented:
Thanks for the updates!  When testing I see two issues:

1) `npm run deploy` isn't working for me.  Does it work for you?  I'm not sure how to interpret this error...
```
$ npm run deploy

> architecture-viewer@1.0.0 deploy /Users/dj/Code/architecture-viewer-swaraj
> gulp deploy

[22:25:53] Using gulpfile ~/Code/architecture-viewer-swaraj/gulpfile.js
[22:25:53] Starting 'deploy'...
[22:25:53] 'deploy' errored after 95 ms
[22:25:53] TypeError: Cannot read property '0' of null
    at Function.module.exports.Commit.actor (/Users/dj/Code/architecture-viewer-swaraj/node_modules/gift/lib/commit.js:145:56)
    at Function.module.exports.Commit.parse_commits (/Users/dj/Code/architecture-viewer-swaraj/node_modules/gift/lib/commit.js:111:21)
    at /Users/dj/Code/architecture-viewer-swaraj/node_modules/gift/lib/commit.js:71:39
    at ChildProcess.exithandler (child_process.js:294:7)
    at ChildProcess.emit (events.js:315:20)
    at ChildProcess.EventEmitter.emit (domain.js:506:15)
    at maybeClose (internal/child_process.js:1021:16)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:286:5)
    at Process.callbackTrampoline (internal/async_hooks.js:120:14)
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! architecture-viewer@1.0.0 deploy: `gulp deploy`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the architecture-viewer@1.0.0 deploy script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/dj/.npm/_logs/2020-08-06T02_25_53_584Z-debug.log
```

2) Running the tests again later outputs some (apparently benign) warnings.  It seems like it might be the gulp step that is copying the `tests` folder into the `.publish` folder (not 100% sure though)?
```
$ npm run test

> architecture-viewer@1.0.0 test /Users/dj/Code/architecture-viewer-swaraj
> jest

jest-haste-map: duplicate manual mock found: browserMocks
  The following files share their name; please delete one of them:
    * <rootDir>/tests/__mocks__/browserMocks.js
    * <rootDir>/.publish/tests/__mocks__/browserMocks.js

jest-haste-map: duplicate manual mock found: fileMocks
  The following files share their name; please delete one of them:
    * <rootDir>/tests/__mocks__/fileMocks.js
    * <rootDir>/.publish/tests/__mocks__/fileMocks.js

jest-haste-map: duplicate manual mock found: setupTests
  The following files share their name; please delete one of them:
    * <rootDir>/tests/__mocks__/setupTests.js
    * <rootDir>/.publish/tests/__mocks__/setupTests.js

 PASS  tests/parser.test.js
 PASS  tests/validator.test.js

Test Suites: 2 passed, 2 total
Tests:       24 passed, 24 total
Snapshots:   0 total
Time:        1.636s
Ran all test suites.
```
It seems like we're just using gulp to copy the `build` folder into the `gh-pages` branch, so if the solution isn't obvious I can explicitly document that step in the README as a workaround for now.

swarajd commented:
I updated the build script to by default use `--no-prerender`. I also replaced gulp with a separate npm package that ostensibly achieves the same result. Finally, I added a bit of jest config to ignore the `.publish` directory to get rid of that warning.
